Timeseries plot style

FrejaTerpPetersen opened the pull request:
An update to the comparer plotter such that it takes a style and a color input. I wanted to also allow for color input, since this is actually the main reason why I started the issue #463: I want to use my own colors to match my other plots. The style input is just a bonus from my point of view :) 

Design decisions:
- The user can only pass one of the two; either color OR style
- If none are passed, the default colors are used. 
- Inputs can be passed as list of strings or as a single string. If a single string is passed but more than one model result is in the comparer, an error is raised. If len(list) < num_models an error is also raised. 
- If len(color) > num_models: the first color is used for the observations. I am unsure whether this is the best way to do it, since there might be cases, where I just have a large list of colors that I pass to all of my plots regardless of the number of models... On the other hand, allowing the user to choose the observation colors is also useful in case of duplicate colors (default obs color is black, and maybe I want one of the lines to be black too?) 


**Example 1: giving a color input**
```python
# Given a comparer:
cmp = create_comparer()
# Pass a color input to the plot.timeseries() method
ax = cmp.plot.timeseries(color=["red", "blue"])
plt.show()
```
This code yields the following plot: 
![image](https://github.com/user-attachments/assets/231fd79f-8609-4a33-b067-6ba611c4f986)

**Example 2: also specifying the color of the observation points**
Above, the default color of the observations is black. If we want them to be e.g. green instead, we pass it as the first color.
```python
# Pass a color input to the plot.timeseries() method
ax = cmp.plot.timeseries(color=["green","red", "blue"])
plt.show()
```
This yields the following plot: 
![image](https://github.com/user-attachments/assets/38c470fa-4250-484a-996e-82dbfcd1240e)

**Example 3: giving a style input**
We can also pass a style input instead
```python
# Pass a color input to the plot.timeseries() method
ax = cmp.plot.timeseries(style=["r-.", "b--"])
plt.show()
```
![image](https://github.com/user-attachments/assets/f6f30317-bca6-40ee-964b-6575f39f7a35)

FrejaTerpPetersen committed:
Style argument to timeseries plot

FrejaTerpPetersen committed:
first value is obs style

FrejaTerpPetersen committed:
Added color argument

FrejaTerpPetersen committed:
Created function for checking inputs

FrejaTerpPetersen committed:
Decided to only allow color OR style input

FrejaTerpPetersen committed:
Wrote tests

FrejaTerpPetersen committed:
Deleted test call

FrejaTerpPetersen committed:
docstring update

FrejaTerpPetersen committed:
Design decision

FrejaTerpPetersen committed:
Removed unused import

FrejaTerpPetersen committed:
Fix type error: color was not subscriptable if None

FrejaTerpPetersen committed:
Wrong type error ... Moved default coloring

FrejaTerpPetersen committed:
Import default colors

FrejaTerpPetersen committed:
Allow color to be tuple

FrejaTerpPetersen committed:
Moved import outside plot code

ecomodeller commented:
@FrejaTerpPetersen for the benefit of the interested reader... could you please include some sample code and screenshots of the new functionality in this description above. Visuals are easy to grasp, when browsing around.

FrejaTerpPetersen commented:
Sure thing - I updated the description

ecomodeller commented on modelskill/comparison/_comparer_plotter.py:
```diff
@@ -79,32 +82,66 @@ def timeseries(
                 mod = cmp.raw_mod_data[key]._values_as_series
-                mod.plot(ax=ax, color=MOD_COLORS[j])
+                if style is not None:
+                    mod.plot(ax=ax, style=style[j])
```
I suggest to add a check (outside the loop) that the length of the style list is the same as the number of models (or models+1 obs, see comment below).

jsmariegaard committed:
Merge remote-tracking branch 'origin/main' into timeseries-plot-style

jsmariegaard commented on modelskill/comparison/_comparer_plotter.py:
```diff
@@ -79,32 +82,66 @@ def timeseries(
+        # Color for observations:
+        obs_color = cmp.data[cmp._obs_name].attrs["color"]
+
+        # if color is None and style is None:  # Use default values for colors
```
remove unused code

jsmariegaard committed:
Example notebook with timeseries plotting

jsmariegaard reviewed (changes requested):
Great work! 

I would like to see a few changes:

- please change _check_kwarg_and_convert_to_list to only check for either style or color and rename to _check_arg_length_match_nmodels(arg, arg_name, n_mod) or to something with _parse (as it returns the same argument again)
- move the function out of _misc.py and to top of _comparer_plotter.py (it is only used in this module)
- I think we should allow either lengths n_mod and n_mod+1 only (i.e. also issue exception if length is too long)
- It seems that the color and style args work a little different if you want to change the obs apperance also. Is it correct that you cannot change the style of the observation.
- I guess the style and color arguments only applied to matplotlib backend and not plotly. If that is the case - it should be clear in the docstring. Better: add it to plotly backend also. 
- currently all kwargs are accepted, e.g. `cmp.plot.timeseries(assdsddsdas="Reds")` does not give me an error message. It should. It is not strictly part of this PR, but as this function is now "open", we should fix it.  
- I have added a short notebook with examples - maybe you could expand it a little bit?

jsmariegaard commented:
closes #463

ecomodeller commented:
@FrejaTerpPetersen What is the status on this PR?

FrejaTerpPetersen committed:
Specify only matplotlib backend in docstring

FrejaTerpPetersen committed:
Moved check function from misc to top of comparer_plotter

FrejaTerpPetersen committed:
Rewrite check function

FrejaTerpPetersen committed:
remove unused code, update test

FrejaTerpPetersen commented on modelskill/comparison/_comparer_plotter.py:
```diff
@@ -79,32 +82,66 @@ def timeseries(
             ax.scatter(
                 cmp.time,
                 cmp.data[cmp._obs_name].values,
                 marker=".",
```
Yes, observations can't have their style changed now. THis is a design choice, I guess? We could allow the user to input style for obs?

ecomodeller replied:
I see the color can be changed. But not the marker.

FrejaTerpPetersen commented:
@ecomodeller I think it is ready for merge, unless you have further comments. 


I updated according to Jespers comments:

- please change _check_kwarg_and_convert_to_list to only check for either style or color and rename to _check_arg_length_match_nmodels(arg, arg_name, n_mod) or to something with _parse (as it returns the same argument again)
       ->I changed the function to your first suggestion. Not familiar with '_parse' - what would that mean?
- move the function out of _misc.py and to top of _comparer_plotter.py (it is only used in this module)
        ->Done
- I think we should allow either lengths n_mod and n_mod+1 only (i.e. also issue exception if length is too long)
       -> Hmmm, my reason for allowing more colors is that the user of the function might just have some generic list of colors from a theme that they want to input without subsetting. But maybe that is an unusual use-case? 
- I guess the style and color arguments only applied to matplotlib backend and not plotly. If that is the case - it should be clear in the docstring. Better: add it to plotly backend also.
       -> I changed docstring - not familiar with plotly so can't quickly do this. 
- currently all kwargs are accepted, e.g. cmp.plot.timeseries(assdsddsdas="Reds") does not give me an error message. It should. It is not strictly part of this PR, but as this function is now "open", we should fix it.
       -> Hmm, not sure I understand exactly ... 


@ecomodeller, @jsmariegaard - do you want to review before merging?

ecomodeller committed:
Not (yet) implemented in plotly

ecomodeller committed:
Add plotly as test dependency

ecomodeller committed:
Merge branch 'main' into timeseries-plot-style

ecomodeller commented on modelskill/comparison/_comparer_plotter.py:
```diff
@@ -31,6 +31,17 @@
+    if isinstance(arg, str):
+        # If arg is str, convert to list (for looping)
+        arg = [arg]
+    if arg is not None and len(arg) < n_mod:
```
This doesn't account for the case where the user passes more arguments than n_mod. This should not be silently ignored.
